=act #20910 optimize ByteString#copyToBuffer

ktoso opened the pull request:
I couldn't resist trimming off that perf penalty we found the other day as it sprang into my face today again.

The case that matters: `1 489 195.631 ops/s => 113 535 003.488 ops/s` 🎉 

Resolves #20910

ktoso committed:
=act #20910 optimize ByteString#copyToBuffer

patriknw commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -147,6 +147,20 @@ object ByteString {
+      writeToBuffer(buffer, offset = 0)
+
+    /** Specialized for internal use, writing multiple ByteString1C into the same ByteBuffer. */
+    def writeToBuffer(buffer: ByteBuffer, offset: Int): Int = {
```
does it need to be public?

ktoso replied:
Just realised they're not needed – cna be done directly in copyToBuffer.

ktoso replied:
No, needed still... checking if needs to be public hm

ktoso replied:
Has to be at least `private[akka]` (makes sense, I'll make it so), since it needs to be called from `ByteStrings`

ktoso commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -452,7 +491,7 @@ sealed abstract class ByteString extends IndexedSeq[Byte] with IndexedSeqOptimiz
    * @return the number of bytes actually copied
    */
-  def copyToBuffer(buffer: ByteBuffer): Int = iterator.copyToBuffer(buffer)
+  def copyToBuffer(buffer: ByteBuffer): Int // specialized in sub-classes
```
realised this is binary incompatible.
Shall I revert and override in subclasses?

Not that anyone should be extending ByteString though hm

patriknw replied:
is this change binary compatible?

patriknw replied:
yeah, better be on the safe side

patriknw commented on akka-actor/src/main/scala/akka/util/ByteString.scala:
```diff
@@ -312,6 +339,18 @@ object ByteString {
+      while (i < bytestrings.length) {
+        val bs: ByteString1 = bytestrings(i)
+        written += bs.writeToBuffer(buffer)
+        if (written < bs.length) return written
```
normally we avoid `return`, such things can be written with tailrec

ktoso replied:
I agree, good idea 👍

ktoso commented:
Addressed feedback, failure was MiMa (which we noticed too, so that won't trigger now as I made it compatible – there is no performance impact on choosing to make it compatible)

ktoso commented:
Failure was unrelated `[error] 	docs.http.javadsl.server.directives.FileUploadDirectivesExamplesTest
`

ktoso commented:
PLS BUILD

ktoso commented:
I'll leave this one for next sprint.

johanandren commented:
LGTM, awesome!

ktoso commented:
I'll measure a bit more so we can have a nice before/after
